### PR TITLE
fix(tarballs): use correct top-level directory name in sdists

### LINKED ISSUE #1315: default_build_sdist produces sdists rooted at build_dir's name instead of {name}-{version}
## Problem

The [sdist specification](https://packaging.python.org/en/latest/specifications/source-distribution-format/#source-distribution-file-format) requires that a source distribution contain exactly one top-level directory named `{name}-{version}`.

When a package sets `build_dir` in its package settings (the monorepo-subdirectory case), `default_build_sdist` produces a tarball rooted at the *build directory's* name instead. The filename is correct; only the directory inside is wrong.

For example, a package `mlserver-xgboost` whose sources live in `runtimes/xgboost/` yields `mlserver-xgboost-1.7.1.tar.gz` that unpacks into `xgboost/`.

## Cause

`default_build_sdist` tars from `build_dir` using the parent as the archive prefix:

https://github.com/python-wheel-build/fromager/blob/4eeb0f415c237578b2b970c516bbf2cee75b409f/src/fromager/sources.py#L549-L553

`tar_reproducible` then computes each entry as `arcname = os.path.relpath(fn, prefix)`:

https://github.com/python-wheel-build/fromager/blob/4eeb0f415c237578b2b970c516bbf2cee75b409f/src/fromager/tarballs.py#L58

So when `build_dir` is `<sdist_root_dir>/runtimes/xgboost`, `prefix` becomes `<sdist_root_dir>/runtimes` and every path is rooted at `xgboost/`. When `build_dir` is unset it equals `sdist_root_dir`, the prefix is its parent, and the root comes out correct — which is why this only affects packages that configure `build_dir`.

## Reproducer

No network or package settings required; it calls `default_build_sdist` directly.

```python
"""Reproducer: sdist top-level directory is build_dir's name, not {name}-{version}."""
import pathlib, tarfile, tempfile, types
from packaging.requirements import Requirement
from packaging.version import Version
from fromager import sources


def build(subdir: str | None) -> list[str]:
    tmp = pathlib.Path(tempfile.mkdtemp())
    sdist_root = tmp / "mypkg-1.0"
    build_dir = sdist_root / subdir if subdir else sdist_root
    build_dir.mkdir(parents=True)
    (build_dir / "setup.py").write_text("from setuptools import setup; setup()\n")
    out = tmp / "sdists-builds"
    out.mkdir()
    sdist = sources.default_build_sdist(
        types.SimpleNamespace(sdists_builds=out),  # ctx
        {},                                        # extra_environ
        Requirement("mypkg"),                      # req
        Version("1.0"),                            # version
        sdist_root,                                # sdist_root_dir
        None,                                      # build_env (unused here)
        build_dir,                                 # build_dir
    )
    with tarfile.open(sdist) as tar:
        return sorted({n.split("/")[0] for n in tar.getnames()})


print(f"build_dir unset (== sdist_root_dir) -> {build(None)}")
print(f"build_dir = 'python'               -> {build('python')}")
print(f"expected in both cases             -> ['mypkg-1.0']")
```

Run with `uv run --with-editable . python repro.py`:

```
build_dir unset (== sdist_root_dir) -> ['mypkg-1.0']
build_dir = 'python'               -> ['python']
expected in both cases             -> ['mypkg-1.0']
```

## Impact

Installers tolerate the layout — `pip` and `uv` build these sdists without complaint — so this does not break installs. It does affect anything that relies on the documented structure:

**Name collisions between unrelated projects.** The root directory is no longer unique per package, so distinct projects can produce sdists that unpack over each other. In one published index built with fromager, four unrelated projects (`ctranslate2`, `flatbuffers`, `tacozip`, `triton`) all emit sdists rooted at `python/`. Extracting two of them into the same directory silently loses one:

```
after flatbuffers-25.12.19.tar.gz: out/python/PKG-INFO says Name: flatbuffers
after ctranslate2-4.8.1.tar.gz:    out/python/PKG-INFO says Name: ctranslate2
what survived: python
```

No error is raised. Workflows that unpack a set of sdists into a common tree — source auditing, license scanning, vendoring, offline mirroring — drop packages here rather than failing loudly.

**Identity mismatch.** `PKG-INFO` inside the archive carries the correct `Name:` while the directory it sits in does not, so tools that infer the project from the directory name disagree with tools that read the metadata.

Same published index: 72 of 3,384 sdists have a top-level directory that is not `{name}-{version}`. Observed roots include `python/`, `xgboost/`, `sklearn/`, `lightgbm/`, `onnx/`, `core/`, `common/`, `tabular/`, `timeseries/`, `ogx_api/`.

This may also matter for #1311, which proposes unpacking the generated sdist and building the wheel from it — that pipeline would have to special-case a root directory that isn't `{name}-{version}`.

## Suggested fix

Root the archive entries at `{name}-{version}` regardless of where `build_dir` points, rather than deriving the prefix from `build_dir.parent`. `tar_reproducible` already takes the prefix as a parameter, so this could be an explicit arcname root rather than a path-relative computation.

### PR DESCRIPTION
**Pull request by jlarkin09**
When a package specifies build_dir in settings (monorepo subdirectory), default_build_sdist was creating tarballs rooted at the build_dir's name instead of {name}-{version} as required by PEP 427.

For example, mlserver-xgboost with build_dir=runtimes/xgboost/ produced mlserver-xgboost-1.7.1.tar.gz unpacking to xgboost/, causing name collisions and identity mismatches.

Changes:
- Add arcname_root parameter to tar_reproducible() to explicitly set the top-level directory name in archives
- Pass normalized {name}-{version} as arcname_root in default_build_sdist()
- Add test to verify correct archive structure with arcname_root

Closes #1315

### CODE REVIEW

File: src/fromager/sources.py [hunk 1]
```diff
@@ -550,6 +551,7 @@ def default_build_sdist(
             tar=sdist,
             basedir=build_dir,
             prefix=build_dir.parent,
+            arcname_root=f"{canonicalize_name(req.name).replace('-', '_')}-{version}",
```
[the remark this answers is not in the document]
Comment (jlarkin09) [reply src/fromager/sources.py]: Fixed in #1328

File: src/fromager/sources.py [hunk 2]
```diff
@@ -533,7 +534,8 @@ def default_build_sdist(
     #
     # For cases where the PEP 517 approach works, use
     # pep517_build_sdist().
-    sdist_filename = ctx.sdists_builds / f"{req.name}-{version}.tar.gz"
+    dist_name = canonicalize_name(req.name).replace("-", "_")
+    sdist_filename = ctx.sdists_builds / f"{dist_name}-{version}.tar.gz"
```
Comment (mprpic) [inline src/fromager/sources.py:538]: This conflicts with the code written in #1328, you might want to stagger the PRs on top of each other.
